src/main/java/com/google/devtools/build/lib/bazel/bzlmod/GsonTypeAdapterUtil.java

```diff
@@ -283,6 +301,68 @@ public Optional<T> read(JsonReader jsonReader) throws IOException {
     }
   }
 
+  /**
+   * Converts Guava tables into a JSON array of 3-tuples (one per cell). Each 3-tuple is a JSON
+   * array itself (rowKey, columnKey, value).
+   */
+  public static final TypeAdapterFactory IMMUTABLE_TABLE =
+      new TypeAdapterFactory() {
+        @Nullable
+        @Override
+        @SuppressWarnings("unchecked")
+        public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> typeToken) {
+          if (typeToken.getRawType() != ImmutableTable.class) {
+            return null;
+          }
+          Type type = typeToken.getType();
+          if (!(type instanceof ParameterizedType)) {
+            return null;
+          }
+          Type[] typeArgs = ((ParameterizedType) typeToken.getType()).getActualTypeArguments();
+          if (typeArgs.length != 3) {
+            return null;
+          }
+          var typeArgAdapters = Arrays.stream(typeArgs)
+              .map(t -> (TypeAdapter<Object>) gson.getAdapter(TypeToken.get(t)))
+              .collect(Collectors.toList());
+          if (typeArgAdapters.contains(null)) {
+            return null;
+          }
+          return (TypeAdapter<T>) new TypeAdapter<ImmutableTable<Object, Object, Object>>() {
+            @Override
+            public void write(JsonWriter jsonWriter, ImmutableTable<Object, Object, Object> t)
+                throws IOException {
+              jsonWriter.beginArray();
+              for (Table.Cell<Object, Object, Object> cell : t.cellSet()) {
+                jsonWriter.beginArray();
```

Comment:
I think this is my confusion with the Javadoc: there aren't any tuples here (are tuples even a json concept?), just an array of arrays.

Reply:
tuples are not JSON concepts. I wanted to emphasize that each "sub-array" has exactly 3 elements, so conceptually a 3-tuple.